### PR TITLE
feat: add duval's algorithm

### PR DESCRIPTION
**Pull request by aminegh20**
#### Description of Change
Added the implementation of Duval's Algorithm for finding the lexicographically smallest cyclic shift of a sequence. The algorithm is based on the concept of Lyndon words and works in O(n) time, making it efficient for this type of task. This implementation is generic and works with various sequence types, such as strings, vectors, arrays, and deques.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: strings/duval.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+ * @author [Amine Ghoussaini](https://github.com/aminegh20)
+*/
+
+#include <array>    ///< for std::array
+#include <cassert>  ///< for assert
+#include <cstddef>  ///< for std::size_t
+#include <deque>    ///< for std::deque
+#include <iostream> ///< for std::cout and std::endl
+#include <string>   ///< for std::string
+#include <vector>   ///< for std::vector
```
Comment (realstealthninja) [inline strings/duval.cpp:35]:
```suggestion
#include <array>    /// for std::array
#include <cassert>  /// for assert
#include <cstddef>  /// for std::size_t
#include <deque>    /// for std::deque
#include <iostream> /// for std::cout and std::endl
#include <string>   /// for std::string
#include <vector>   /// for std::vector
```

File: strings/duval.cpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
```
Comment (realstealthninja) [inline strings/duval.cpp:36]:
```suggestion
/**
 * @brief string manipulation algorithms
 * @namespace
 */
```

File: strings/duval.cpp [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+ * @returns the 0-indexed position of the least cyclic shift of the sequence
+ */
+template <typename T>
+size_t duval(const T& s) {
+    size_t n = s.size();
+    size_t i = 0, ans = 0;
+    while (i < n) {
+        ans = i;
+        size_t j = i + 1, k = i;
+        while (j < n + n && s[j % n] >= s[k % n]) {
```
Comment (realstealthninja) [inline strings/duval.cpp:51]:
```suggestion
        while (j < (n + n) && s[j % n] >= s[k % n]) {
```
For clarity

File: strings/duval.cpp [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+    // Test 1
+    std::string s1 = "abcab";
+    assert(duval(s1) == 3);
+
+    // Test 2
+    std::string s2 = "011100";
+    assert(duval(s2) == 4);
+
+    // Test 3
```
Comment (realstealthninja) [inline strings/duval.cpp:80]: Add a case for an empty array

File: strings/duval.cpp [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+#include <vector>   ///< for std::vector
+
+namespace string {
+/**
+ * @brief Find the lexicographically smallest cyclic shift of a sequence.
+ * @tparam T type of the sequence
+ * @param s the sequence
+ * @returns the 0-indexed position of the least cyclic shift of the sequence
+ */
+template <typename T>
```
Comment (realstealthninja) [inline strings/duval.cpp:44]: Unsure if this needs to be this generic
Why not just a basic string? std::basic_string? The algorithm is entirely about strings right?

File: strings/duval.cpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    std::array<int, 5> a = {1, 2, 3, 4, 5};
+    assert(duval(a) == 0);
+
+    // Test 5
+    std::deque<char> d = {'a', 'z', 'c', 'a', 'b'};
+    assert(duval(d) == 3);
+
+    std::cout << "All tests passed!" << std::endl;
+}
+
```
Comment (realstealthninja) [inline strings/duval.cpp:94]:
```suggestion

/**
 * @brief main function
 * @returns 0 on exit
 */
```

File: strings/duval.cpp [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+int main() {
+    test();
```
Comment (realstealthninja) [inline strings/duval.cpp:96]:
```suggestion
    test(); // run self test implementations
```

File: strings/duval.cpp [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        while (i <= k) {
+            i += j - k;
+        }
+    }
+    return ans;
+    // returns 0-indexed position of the least cyclic shift
+}
+
+}  // namespace string
+
```
Comment (realstealthninja) [inline strings/duval.cpp:68]:
```suggestion
/**
 * @brief self test implementation
 * returns void
 */
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: strings/duval.cpp (hunk 5 above)
Comment (aminegh20) [reply strings/duval.cpp:44]: The point here is that you could pass in an array of integers as well. In theory, this algorithm is is studied in the context of strings, but the idea here is that you do not need to have "chars". In fact, you could even pass in an array of custom comparable objects.

File: strings/duval.cpp [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    assert(duval(s1) == 3);
+
+    // Test 2
+    std::string s2 = "011100";
+    assert(duval(s2) == 4);
+
+    // Test 3
+    std::vector<int> v = {5, 2, 1, 3, 4};
+    assert(duval(v) == 2);
+
```
Comment (realstealthninja) [inline strings/duval.cpp:91]: add a test case with a negative number aswell

**Review by realstealthninja — CHANGES_REQUESTED** [review]

**Review by realstealthninja — APPROVED** [review]
Looks good! thank you for this contribution

File: strings/duval.cpp [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+ * @param s the sequence
+ * @returns the 0-indexed position of the least cyclic shift of the sequence
+ */
+template <typename T>
+size_t duval(const T& s) {
+    size_t n = s.size();
+    size_t i = 0, ans = 0;
+    while (i < n) {
+        ans = i;
+        size_t j = i + 1, k = i;
```
Comment (realstealthninja) [inline strings/duval.cpp:54]:
```suggestion
std::size_t duval(const T& s) {
    std::size_t n = s.size();
    std::size_t i = 0, ans = 0;
    while (i < n) {
        ans = i;
        std::size_t j = i + 1, k = i;
```

**Review by realstealthninja — COMMENTED** [review]
i would recommend adding std:: prefixes.
